core/bin/snapshots_creator/src/main.rs

```diff
@@ -0,0 +1,251 @@
+    let output_filepath_prefix = blob_store.get_storage_prefix::<SnapshotFactoryDependencies>();
+    let output_filepath = format!("{output_filepath_prefix}/{filename}");
+    tracing::info!(
+        "Finished processing factory dependencies, output stored in {}",
+        output_filepath
+    );
+    Ok(output_filepath)
+}
+
+async fn run(blob_store: Box<dyn ObjectStore>, pool: ConnectionPool) -> anyhow::Result<()> {
+    let mut conn = pool.access_storage().await?;
+    let start_time = seconds_since_epoch();
+
+    let l1_batch_number = conn.blocks_dal().get_sealed_l1_batch_number().await? - 1; // we subtract 1 so that after restore, EN node has at least one l1 batch to fetch
+
+    if conn
+        .snapshots_dal()
+        .get_snapshot_metadata(l1_batch_number)
+        .await?
+        .is_some()
+    {
+        tracing::info!(
+            "Snapshot for L1 batch number {} already exists, exiting",
+            l1_batch_number
+        );
+        return Ok(());
+    }
+
+    let miniblock_number = conn
+        .blocks_dal()
+        .get_miniblock_range_of_l1_batch(l1_batch_number)
+        .await?
+        .unwrap()
+        .1;
+    let storage_logs_count = conn
+        .snapshots_creator_dal()
+        .get_storage_logs_count(l1_batch_number)
+        .await?;
+
+    drop(conn);
+
+    //TODO load this from config
+    let chunk_size = 1_000_000;
+    let chunks_count = ceil_div(storage_logs_count, chunk_size);
+
+    tracing::info!(
+        "Creating snapshot for storage logs up to miniblock {}, l1_batch {}",
+        miniblock_number,
+        l1_batch_number.0
+    );
+    tracing::info!(
+        "Starting to generate {} chunks of max size {}",
+        chunks_count,
+        chunk_size
+    );
+
+    let factory_deps_output_file =
+        process_factory_deps(&*blob_store, &pool, miniblock_number, l1_batch_number).await?;
+
+    let mut storage_logs_output_files = vec![];
+
+    for chunk_id in 0..chunks_count {
+        tracing::info!(
+            "Processing storage logs chunk {}/{chunks_count}",
+            chunk_id + 1
+        );
+        let output_file = process_storage_logs_single_chunk(
+            &*blob_store,
+            &pool,
+            l1_batch_number,
+            chunk_id,
+            chunk_size,
+            chunks_count,
+        )
+        .await?;
+        storage_logs_output_files.push(output_file.clone());
+    }
+
+    let mut conn = pool.access_storage().await?;
+
+    conn.snapshots_dal()
+        .add_snapshot(
+            l1_batch_number,
+            &storage_logs_output_files,
+            factory_deps_output_file,
+        )
+        .await?;
+
+    METRICS.snapshot_l1_batch.set(l1_batch_number.0.as_u64());
+    METRICS.storage_logs_chunks_count.set(chunks_count);
+    METRICS
+        .snapshot_generation_timestamp
+        .set(seconds_since_epoch());
+    METRICS
+        .snapshot_generation_duration
+        .set(seconds_since_epoch() - start_time);
```

Comment:
IMO, it would make more sense to track processing of a single chunk, so that this metric is more "dynamic" and allows to measure the health / speed of snapshot creation while it's in progress. Optionally, you can split it by stages (loading from Postgres, saving to GCS for now; saving to Postgres when fault-tolerance is implemented). It would be a `Histogram<Duration>`; see other metric definitions for the examples of usage, such as [in `MetadataCalculator`](https://github.com/matter-labs/zksync-era/blob/38bb4823c7b5e0e651d9f531feede66c24afd19f/core/lib/zksync_core/src/metadata_calculator/metrics.rs#L114).

Reply:
I was thinking a quite a lot about it, but so far I don't yet see, how our operations could benefit from it. I would mostly look on total snapshot duration to spot regressions. I don't imagine looking at it during normal operation and if some particular snapshot generation fails / is very slow, the logs would be the first place to diagnose what has happened and should contain all needed info about speed of snapshot creation. Why do you think it would be useful to have those details directly in metrics?

Reply:
It would allow to track that a particular snapshot generation is slow (or more generally, will allow to track its progress). I may be missing something, but how would it be tracked otherwise using metrics? I want to have this info on a Grafana dashboard to be able to assess snapshot generation health just by looking at; additionally, metrics could be used for altering. As another use case: measure the impact of integrations (GCS / Postgres split) depending on their configuration or code changes.

Reply:
I added some additional fine-grained metrics